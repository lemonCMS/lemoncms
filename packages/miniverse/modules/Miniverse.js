import {
  concat,
  ReplaySubject,
  zip,
} from 'rxjs';
import { ApiService } from './ApiService';
import { EventService } from './EventService';

class Miniverse {
  /**
   *
   * @type {{}}
   */
  services = {};

  /**
   *
   * @type concat
   * */
  concat = concat;

  /**
   *
   * @type {{}}
   */
  helpers = {};

  /**
   *
   * @type {ApiService}
   */
  apiService = null;

  /**
   *
   * @type {EventService}
   */
  eventService = null;

  /**
   *
   * @param services
   * @param apiService
   */
  constructor(services, apiService = null) {

    if (apiService === null) {
      this.apiService = new ApiService()
    } else {
      this.apiService = apiService;
    }

    this.eventService = new EventService;

    this.services = this.objectMap(services, Service => {
      const service = new Service(this, this.apiService, this.eventService);
      if (typeof service.init === 'function') {
        service.setConfCallback(service.init());
      }
      return service;
    });
  }

  /**
   *
   * @param name
   * @returns {boolean}
   */
  isService(name) {
    return typeof this.services[name] !== 'undefined';
  }

  /**
   *
   * @returns {*}
   */
  eject = () => {
    const subjects = {};
    Object.keys(this.services).forEach(service => {
      subjects[service] = this.services[service].eject();
    });
    return this._waitAllDone(subjects);
  };

  /**
   *
   * @returns {*}
   */
  insert = (data) => {
    const subjects = {};
    Object.keys(data).forEach(service => {
      if (this.isService(service)) {
        subjects[service] = this.services[service].insert(data[service]);
      }
    });
    return this._waitAllDone(subjects);
  };

  /**
   *
   * @param name
   * @returns {boolean|*}
   */
  getService = name => {
    if (typeof this.services[name] === 'undefined') {
      throw Error(`Service ${name} is not registered in services`);
    }
    return this.services[name];
  };

  /**
   *
   * @returns {EventService}
   */
  getEventService() {
    return this.eventService;
  }

  /**
   *
   * @param helpers
   */
  setHelpers = helpers => {
    const { client, ...rest } = helpers;
    this.client = client;
    this.helpers = rest;
  };

  /**
   *
   * @param name
   * @returns {boolean|*}
   */
  getHelper = name => {
    if (typeof this.helpers[name] === 'undefined') {
      console.error(`Helper ${name} is not registered in helpers`);
      return false;
    }
    return this.helpers[name];
  };

  /**
   *
   * @param object
   * @param mapFn
   * @returns {{}}
   */
  objectMap = (object, mapFn) => {
    return Object.keys(object).reduce((result, key) => {
      result[key] = mapFn(object[key]);
      return result;
    }, {});
  };

  _waitAllDone = subjects => {
    const subject = new ReplaySubject();
    const toExport = Object.keys(subjects);
    let nextAfter = toExport.length;
    const preparedExport = {};

    if (nextAfter === 0) {
      subject.next({});
      subject.complete();
    }

    toExport.forEach(key => {
      subjects[key].subscribe(data => {
        preparedExport[key] = data;
        nextAfter -= 1;
        if (nextAfter <= 0) {
          subject.next(preparedExport);
          subject.complete();
        }
      });
    });
    return subject;
  };

  zip(...rest) {
    const toZip = rest.map(apiCall => {
      if (typeof apiCall === 'object' && typeof apiCall.subject === 'function') {
        return apiCall.subject();
      }
      return apiCall;
    });
    return zip(...toZip);
  }
}

export default Miniverse;
