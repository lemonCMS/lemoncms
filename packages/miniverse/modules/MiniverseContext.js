import { createContext } from 'react';

const MiniverseContext = createContext({});

























export default MiniverseContext;


