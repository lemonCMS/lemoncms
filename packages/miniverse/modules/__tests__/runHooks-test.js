import React from 'react';
import renderer from 'react-test-renderer';
import RunHooks from '../runHooks';
import Miniverse from '../Miniverse';
import Github from './__mock__/Github-mock';
import Component from './components/DecoratedComponent';
import ComponentFakeHook from './components/ComponentFakeHook';
import ComponentThatDoesToMuch from './components/ComponentThatDoesToMuch';
import ComponentWithError from './components/ComponentWithError';
import WatchComponent from './components/WatchComponent';
import PlainComponent from './components/PlainComponent';
import MiniverseContext from '../MiniverseContext';
import { ajax } from 'rxjs/ajax';
import { of } from 'rxjs';

jest.mock("tiny-warning");
jest.mock("rxjs/ajax");


describe('RunHooks should do it', () => {
  test('Should retrieve data', (done) => {
    const initiatedServiceProvider = new Miniverse({
      Github
    });
    const providers = {
      miniverse: initiatedServiceProvider
    };

    const components = [];
    components.push(Component);
    ajax.mockImplementationOnce(() =>
      of({
        response: { results: ["cat.jpg"] }
      })
    );
    const rick = new RunHooks('fetch', components, providers);
    rick.subscribe(() => {
      done();


      process.nextTick(() => {
        initiatedServiceProvider.eject().subscribe((data) => {
          expect(data.Github['111578632']).toBeDefined();
          done();
        });
      });
    });
  });

  test('Should not return data', (done) => {
    const initiatedServiceProvider = new Miniverse({
      Github
    });
    const providers = {
      miniverse: initiatedServiceProvider
    };

    const components = [];
    components.push(ComponentFakeHook);
    ajax.mockImplementationOnce(() =>
      of({
        response: { results: ["cat.jpg"] }
      })
    );

    expect(() => {
      const rick = new RunHooks('fetch', components, providers);
      rick.subscribe(() => {
        process.nextTick(() => {

          initiatedServiceProvider.eject().subscribe((data) => {
            expect(data.Github['111578632']).not.toBeDefined();
          });
        });
      });
    }).toThrowError();
    done();

  });

  test('Should just return no execution', (done) => {
    const initiatedServiceProvider = new Miniverse({
      Github
    });
    const providers = {
      miniverse: initiatedServiceProvider
    };

    const components = [];
    components.push(PlainComponent);
    const rick = new RunHooks('fetch', components, providers);
    rick.subscribe(() => {
      initiatedServiceProvider.eject().subscribe((data) => {
        expect(data.Github['111578632']).not.toBeDefined();
        done();
      });
    });
  });

  /*test('Should throw warning about to many renders', (done) => {
    const initiatedServiceProvider = new Miniverse({
      Github
    });
    const providers = {
      miniverse: initiatedServiceProvider
    };

    const components = [];
    components.push(ComponentThatDoesToMuch);
    components.push(ComponentThatDoesToMuch);
    ajax.mockImplementationOnce(() =>
      of({
        response: { results: ["cat.jpg"] }
      })
    );
    const rick = new RunHooks('fetch', components, providers);
    rick.subscribe(() => {
      process.nextTick(() => {
        initiatedServiceProvider.eject().subscribe(() => {
          expect(mockTinyWarning).toBeCalled();
          done();
        });
      });
    });
  });*/

  test('Should resolve a error with multiple componentns', (done) => {
    const initiatedServiceProvider = new Miniverse({
      Github
    });
    const providers = {
      miniverse: initiatedServiceProvider
    };

    const components = [];
    components.push(ComponentWithError);
    components.push(ComponentThatDoesToMuch);
    ajax.mockImplementationOnce(() =>
      of({
        response: { results: ["cat.jpg"] }
      })
    );
    const rick = new RunHooks('fetch', components, providers);
    rick.subscribe(() => {

    }, (error) => {
      process.nextTick(() => {
        expect(error).toBeDefined()
        done();
      });
    });
  })

  test('Should resolve a error with single componentns', (done) => {
    const initiatedServiceProvider = new Miniverse({
      Github
    });
    const providers = {
      miniverse: initiatedServiceProvider
    };

    const components = [];
    components.push(ComponentWithError);
    ajax.mockImplementationOnce(() =>
      of({
        response: { results: ["cat.jpg"] }
      })
    );
    const rick = new RunHooks('fetch', components, providers);
    rick.subscribe(() => {

    }, (error) => {
      process.nextTick(() => {
        expect(error).toBeDefined()
        done();
      });
    });
  })

  test('Watch Component should resolve data', (done) => {
    const initiatedServiceProvider = new Miniverse({
      Github
    });
    const providers = {
      miniverse: initiatedServiceProvider
    };

    const components = [];
    components.push(Component);
    components.push(WatchComponent);
    ajax.mockImplementationOnce(() =>
      of({
        response: { results: ["cat.jpg"] }
      })
    );
    const rick = new RunHooks('fetch', components, providers);
    rick.subscribe(() => {
      const tree = renderer
        .create(<MiniverseContext.Provider value={providers}>
          <WatchComponent />
        </MiniverseContext.Provider>);
      process.nextTick(() => {
        const output = tree.toJSON();
        expect(output).toMatchSnapshot();
        done();
      });
    });
  })
})
;

