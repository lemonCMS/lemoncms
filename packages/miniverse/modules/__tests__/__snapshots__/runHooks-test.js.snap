// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunHooks should do it Watch Component should resolve data 1`] = `
<div>
  Rerenderd: 
  1
   
  {"users":{"results":["cat.jpg"]}}
</div>
`;

exports[`RunHooks should do it Watch Component should resolve data 2`] = `
<div>
  Rerenderd: 
  2
   
  {"users":{"results":["cat.jpg"]}}
</div>
`;
